### ISSUE TITLE
UID/GID not set to zero for all files added with ADD command

### ISSUE THREAD

**Issue by JonathonReinhart** [docker/docker#22056]
This is a continuation of #3950, which was supposedly fixed by #3971. It does not appear to be completely fixed, as of Docker 1.9.1 (a34a1d5).

I'm adding three files (owned by 1001:1001) like this in my Dockerfile:
```
ADD subdir/file1 subdir/file2 subdir/file3 /root/.somewhere
```

Result:
```
# ls al /root/.somewhere
-rw-r--r-- 1 1001    1001   100 Apr 14 21:35 file1
-rw-r--r-- 1 root    root   100 Apr 14 21:35 file2
-rw-r--r-- 1 root    root   100 Apr 14 21:35 file3
```

The UID/GID are being set for the all of the files *except* the first.

**Comment by HackToday** [comment]
I used 1.11 docker and not found this issue

```
ubuntu@interlock:~/tetwok$ docker build -t test  .
Sending build context to Docker daemon 4.608 kB
Step 1 : FROM busybox
 ---> 47bcc53f74dc
Step 2 : ADD subdir/file1 subdir/file2 subdir/file3 /root/.somewhere/
 ---> 6402e6fc6bc7
Removing intermediate container 8619da8e2f39
Successfully built 6402e6fc6bc7
ubuntu@interlock:~/tetwok$ docker images
REPOSITORY           TAG                 IMAGE ID            CREATED             SIZE
test                 latest              6402e6fc6bc7        4 seconds ago       1.113 MB
nginx                latest              eb4a127a1188        9 days ago          182.7 MB
ehazlett/interlock   1.0.1               7ec5cfc363b2        3 weeks ago         20.9 MB
busybox              latest              47bcc53f74dc        3 weeks ago         1.113 MB
ubuntu@interlock:~/tetwok$ docker run -it test sh
/ # ls -l /root/.somewhere/file
file1  file2  file3
/ # ls -l /root/.somewhere/
total 4
-rw-rw-r--    1 root     root             6 Apr 15 01:00 file1
-rw-rw-r--    1 root     root             0 Apr 15 01:00 file2
-rw-rw-r--    1 root     root             0 Apr 15 01:00 file3
/ # exit
ubuntu@interlock:~/tetwok$ ls -l 
Dockerfile  subdir/     
ubuntu@interlock:~/tetwok$ ls -l 
Dockerfile  subdir/     
ubuntu@interlock:~/tetwok$ ls -l subdir/
total 4
-rw-rw-r-- 1 ubuntu ubuntu 6 Apr 15 01:00 file1
-rw-rw-r-- 1 ubuntu ubuntu 0 Apr 15 01:00 file2
-rw-rw-r-- 1 ubuntu ubuntu 0 Apr 15 01:00 file3
```

**Comment by JonathonReinhart** [comment]
Same thing on `1.8.2-fc22`:

```
[jreinhart@jreinhart-nuc docker_22056]$ ls -l subdir/
total 8
-rw-rw-r--. 1 jreinhart jreinhart 0 Apr 15 08:14 file1
-rw-rw-r--. 1 jreinhart jreinhart 5 Apr 15 08:15 file2
-rw-rw-r--. 1 jreinhart jreinhart 9 Apr 15 08:15 file3
[jreinhart@jreinhart-nuc docker_22056]$ docker build -t test .
Sending build context to Docker daemon  5.12 kB
Step 0 : FROM debian:8.2
 ---> d4b2ba78e3b4
Step 1 : ADD subdir/file1 subdir/file2 subdir/file3 /root/.somewhere/
 ---> 10f6dd22eea1
Removing intermediate container 8c83082c383e
Successfully built 10f6dd22eea1
[jreinhart@jreinhart-nuc docker_22056]$ docker run --rm -it test  ls -l /root/.somewhere
total 8
-rw-rw-r--. 1 1000 1000 0 Apr 15 12:14 file1
-rw-rw-r--. 1 root root 5 Apr 15 12:15 file2
-rw-rw-r--. 1 root root 9 Apr 15 12:15 file3
[jreinhart@jreinhart-nuc docker_22056]$ docker version
Client:
 Version:      1.8.2-fc22
 API version:  1.20
 Package Version: docker-1.8.2-7.gitcb216be.fc22.x86_64
 Go version:   go1.5.1
 Git commit:   cb216be/1.8.2
 Built:        
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.2-fc22
 API version:  1.20
 Package Version: 
 Go version:   go1.5.1
 Git commit:   cb216be/1.8.2
 Built:        
 OS/Arch:      linux/amd64
```

**Comment by JonathonReinhart** [comment]
Ping @crosbymichael @creack

**Comment by thaJeztah** [comment]
(btw: it looks like you are running a Red Hat build of Docker; https://github.com/projectatomic/docker/commit/cb216be, which contains various modifications, and can in some cases make a difference. Support for those versions should be handled through https://bugzilla.redhat.com)


I was able to reproduce this on 1.8.2, and 1.9.1 on Fedora 22;

```
Client:
 Version:      1.8.2
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   0a8c2e3
 Built:        Wed Oct  7 17:42:41 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.2
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   0a8c2e3
 Built:        Wed Oct  7 17:42:41 UTC 2015
 OS/Arch:      linux/amd64
```

And 

```
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:34:39 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:34:39 UTC 2015
 OS/Arch:      linux/amd64
```

But it looks to be resolved in 1.10.3;

```
Client:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:45:58 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:45:58 2016
 OS/Arch:      linux/amd64
```

Files;

```
ls -la subdir/
total 8
drwxr-xr-x. 2 1001 1001 4096 Apr 15 13:21 .
dr-xr-x---. 6 root root 4096 Apr 15 13:21 ..
-rw-r--r--. 1 1001 1001    0 Apr 15 13:21 file1
-rw-r--r--. 1 1001 1001    0 Apr 15 13:21 file2
-rw-r--r--. 1 1001 1001    0 Apr 15 13:21 file3
```


Dockerfile:
```
FROM alpine
ADD subdir/file1 subdir/file2 subdir/file3 /root/.somewhere/
COPY subdir/file1 subdir/file2 subdir/file3 /root/.somewhere-else/
CMD sh
```

```
docker build -t owner .
...

docker run -it --rm owner

/ # ls -la /root/.somewhere/
total 8
drwxr-xr-x    2 root     root          4096 Apr 15 12:57 .
drwx------    4 root     root          4096 Apr 15 12:58 ..
-rw-r--r--    1 root     root             0 Apr 15 12:53 file1
-rw-r--r--    1 root     root             0 Apr 15 12:53 file2
-rw-r--r--    1 root     root             0 Apr 15 12:53 file3


/ # ls -la /root/.somewhere-else/
total 8
drwxr-xr-x    2 root     root          4096 Apr 15 12:57 .
drwx------    4 root     root          4096 Apr 15 12:58 ..
-rw-r--r--    1 root     root             0 Apr 15 12:53 file1
-rw-r--r--    1 root     root             0 Apr 15 12:53 file2
-rw-r--r--    1 root     root             0 Apr 15 12:53 file3
```

Are you still able to reproduce this on docker 1.10.3 or 1.11.0 ?

**Comment by JonathonReinhart** [comment]
> (btw: it looks like you are running a Red Hat build of Docker; projectatomic@cb216be, which contains various modifications, and can in some cases make a difference. Support for those versions should be handled through https://bugzilla.redhat.com)

@thaJeztah The original bug report was for Docker 1.9.1 (a34a1d5), a non Red Hat build. My comment was regarding a sanity check on a second machine which was using the RH version.

I'm currently unable to update to a different version on the machine running the RH version, because I can't uninstall docker due to [Bug 1246826 - Removing docker-selinux removes selinux-policy](https://bugzilla.redhat.com/show_bug.cgi?id=1246826).

**Comment by thaJeztah** [comment]
@JonathonReinhart alright, let me close this issue, because this should be resolved, but let us know if it isn't resolved after updating to docker 1.10 or 1.11

**Comment by JonathonReinhart** [comment]
Okay, I updated to Docker 1.11.0, and the issue appears to be resolved. Thanks. If I find the MR that fixed it, I will link it here, because the only things I can find about this problem are before it was first the fixed time.

**Comment by JonathonReinhart** [comment]
This was re-fixed in PR #18158; no associated issue.

**Comment by thaJeztah** [comment]
Thanks! Ah, yes, I remember that one :+1:
